The big storage refactoring

scholzj opened the pull request:
### Type of change

- Enhancement / new feature
- Refactoring

### Description

This PR does the refactoring of the storage handling in CO to give us more flexibility for other changes such as storage resizing, adding / removing volumes in JBOD etc. It does following main changes:
* Reworked reconciliation of PVCs and their cleanup (our method during scaledown, garbage collection after cluster deletion)
* Possibility to _replace_ Statefulset after changes to immutable fields (non-cascading delete + create)
* Improvements to Pod cleanup to handle race conditions between PVC deletion and pod restart
* Storing old storage configuration as JSON in annotation to make it easy to revert changes which are not allowed.
* Removal of the old code for deleting clusters (replaced by garbage collection)

+ probably something ore what was standing in the way.

### Checklist

- [ ] Write tests
- [ ] Make sure all tests pass
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally

scholzj commented:
@tombentley @ppatierno It is a draft PR only. But the code should be _final_ and it is tested live in OpenShift. Before I invest into tests, could you please review this to make sure I don't waste time on writing tests for code you will not approve?

scholzj commented:
Anything @tombentley and @ppatierno?

ppatierno commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/KafkaSetOperatorTest.java:
```diff
@@ -134,7 +130,7 @@ public void testNeedsRollingUpdateEnvSomeOtherThing() {
     }
 
-    @Test
+    /*@Test
```
this test is commented out, there is something more to do for having it running or just forgot to remove the comment block?

scholzj replied:
As I said, the tests are not done for this. I wanted to get your comments to the code before writing the tests.

tombentley replied:
Is it still the case that the tests are not done?

scholzj replied:
Yeah, I forgot to delete this test :-). The reverting of the storage changes is not done here anymore - there are new tests in the model classes..

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java:
```diff
@@ -277,53 +276,81 @@ public static String clientsCaCertSecretName(String cluster) {
+        return fromCrd(kafkaAssembly, versions, null);
+    }
+
+    public static KafkaCluster fromCrd(Kafka kafkaAssembly, KafkaVersion.Lookup versions, Storage oldStorage) {
```
just an idea, instead of passing a specific thing like the `Storage` got from the STS in the KAO, could it be useful passing the STS itself for getting the old storage here? Maybe having the current STS instance could be useful in the future? Or maybe I am just overthinking it ...

scholzj replied:
Do you see anything like that in the near future? TBH, I think it is better to keep this logic outside to make it easier to change and keep the `KafkaCluster` and `ZookeeperCluster` cleaner.

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java:
```diff
@@ -1047,7 +1076,10 @@ ReconciliationState createReconciliationState(Reconciliation reconciliation, Kaf
                 future -> {
                     try {
-                        this.kafkaCluster = KafkaCluster.fromCrd(kafkaAssembly, versions);
+                        StatefulSet sts = kafkaSetOperations.get(namespace, KafkaCluster.kafkaClusterName(name));
```
This is actually in execute blocking, so I think this is ok.

scholzj commented:
@strimzi-ci run tests tags=acceptance,regression

scholzj commented:
@ppatierno @tombentley I incorporated your previous comments and fixed existing / added new tests. So this should be hopefully ready for the final review. Could you please have a look?

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java:
```diff
@@ -582,17 +605,47 @@ protected ServicePort createServicePort(String name, int port, int targetPort, I
 
+    protected PersistentVolumeClaim createPersistentVolumeClaimTemplate(String name, PersistentClaimStorage storage) {
+        Map<String, Quantity> requests = new HashMap<>();
+        requests.put("storage", new Quantity(storage.getSize(), null));
```
Might there be other requests which users might want to configure on the PVCs?

scholzj replied:
There are apparently. It seems that for example IBM storage supports IOPS configuration next to the size (whereas in AWS this is for example done through storage class and not PVC). 

But:
* The actual fields are not specified anywhere (It is using the same resources object as are used for CPU and memory). So we do not know what to support.
* Our current CRDs are not using the resources object which would give us the flexibility but instead have just a single specific field for size. This is where this code comes from.

This is not really something what I changed somehow in this PR and I would leave that for another PR. We should probably think whether we should move to using the Resources in our CRDs here as well as part of v1beta1. Another think is that also ephemeral storage might now have a size limit - that is something we might want to have a look at as well in the future.

scholzj replied:
FYI: I opened issue #1504 for this.

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java:
```diff
@@ -98,6 +100,7 @@
     public static final String ENV_VAR_STRIMZI_GC_LOG_ENABLED = "STRIMZI_GC_LOG_ENABLED";
 
     public static final String ANNO_STRIMZI_IO_DELETE_CLAIM = Annotations.STRIMZI_DOMAIN + "/delete-claim";
+    public static final String ANNO_STRIMZI_IO_STORAGE = Annotations.STRIMZI_DOMAIN + "/storage";
```
```suggestion
    /** Annotation on PVCs storing the original configuration (so we can revert changes). */
    public static final String ANNO_STRIMZI_IO_STORAGE = Annotations.STRIMZI_DOMAIN + "/storage";
```

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java:
```diff
@@ -277,53 +276,81 @@ public static String clientsCaCertSecretName(String cluster) {
         result.setJvmOptions(kafkaClusterSpec.getJvmOptions());
+
         result.setConfiguration(new KafkaConfiguration(kafkaClusterSpec.getConfig().entrySet()));
+
```
I never knew you had such strong feelings about blank lines :open_mouth:

scholzj replied:
It makes it a bit more readable for me. I can revert it if you want ;-).

tombentley replied:
FTR MHO is blank lines to separate/group/delineate important bits of code are useful. Blank lines between every statement is harmful because it means I can literally only get half as much code on my screen. Sometimes it is _really helpful_ to have as much code on screen as possible. Also if every statement is separated by a blank line then I suddenly need _two_ blank lines to serve the legitimate separate/group/delineate need, and it just starts getting out of hand.

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java:
```diff
@@ -742,12 +773,43 @@ private void setDataVolumesClaimsAndMountPaths(Storage storage) {
+        if (storage != null) {
+            if (storage instanceof PersistentClaimStorage) {
+                Integer id = ((PersistentClaimStorage) storage).getId();
+                String pvcName = ModelUtils.getVolumePrefix(id) + "-" + name;
```
`pvcBaseName`, because the actual name has the replica number appended?

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/StorageDiff.java:
```diff
@@ -0,0 +1,92 @@
+        this.changesSize = changesSize;
+    }
+
+    private JsonNode getFromPath(Storage current, String pathValue) {
```
I think since you originally copied this from the SSDiff I have changed that method with some slightly nicer code. We should factor out a common method. If you were feeling more ambitious perhaps there's a common base class for these different kinds of diff.

scholzj replied:
I created a simple `AbstractResourceDiff` class. There is not so much they have in common as most of the logic in the constructor is slightly different (log messages, the things we are looking for etc.). But it should help to make it more clear that these are doing the same thing and help to keep them in sync in the future.

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java:
```diff
@@ -1520,27 +1530,262 @@ ReconciliationState createReconciliationState(Reconciliation reconciliation, Kaf
+         * and desired PVCs to be able to delete and recreate them.
+         *
+         * This method exists with the first Pod it finds where deletion is required. This is intentional - if multiple
+         * pods are marked for deletion, the next one will be deleted in the next loop.
```
Why do we do it like this? You should explain.

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java:
```diff
@@ -1520,27 +1530,262 @@ ReconciliationState createReconciliationState(Reconciliation reconciliation, Kaf
+         * @param futureExistingPvcs    Future which will return a list of PVCs which actually exist
+         * @return
+         */
+        Future<Void> maybeCleanPodAndPvc(StatefulSet sts, List<PersistentVolumeClaim> desiredPvcs, Future<List<PersistentVolumeClaim>> futureExistingPvcs)  {
```
`existingPvcsFuture` would be a better name

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java:
```diff
@@ -1520,27 +1530,262 @@ ReconciliationState createReconciliationState(Reconciliation reconciliation, Kaf
+
+            for (PersistentVolumeClaim pvc : deletePvcs)    {
+                String pvcName = pvc.getMetadata().getName();
+                log.debug("{}: Deleting PVC {} for Pod {} based on annotation", reconciliation, pvcName, podName);
```
Include annotation name

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java:
```diff
@@ -1520,27 +1530,262 @@ ReconciliationState createReconciliationState(Reconciliation reconciliation, Kaf
+
+                    persistentClaimDeletion(maybeDeletePvcs, desiredPvcs).setHandler(futureResult.completer());
+                } else {
+                    futurePvcs.fail(res.cause());
```
I think you mean to fail `futureResult` here, not `futurePvcs` which must have already been completed.

If the `res` succeeded but had null `result` theb we'd fail `futurePvcs` here with a null cause. I suspect that's not really what you want. I don't think `res.result()` can actually be null, so probably all you need to do it remove the `&& res.result() != null` from the `if` clause.

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java:
```diff
@@ -1520,27 +1530,262 @@ ReconciliationState createReconciliationState(Reconciliation reconciliation, Kaf
+
+                    persistentClaimDeletion(maybeDeletePvcs, desiredPvcs).setHandler(futureResult.completer());
+                } else {
+                    futurePvcs.fail(res.cause());
```
Same comment here.

tombentley commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaClusterTest.java:
```diff
@@ -1494,4 +1501,191 @@ public void testGetExternalAdvertisedUrlWithoutOverrides() {
+
+        /**********
+         * Persistent storage with claim deletion
+         */
```
I'm guilty of writing this kind of test myself, but I think it would be better to have a different test method for each of these cases. It makes the tests a lot clearer when future refactoring causes some of them to fail.

scholzj replied:
Changed. Honestly, I was my self a bit undecided about this when writing the tests,

tombentley commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ZookeeperClusterTest.java:
```diff
@@ -659,4 +663,115 @@ public void testNetworkPolicy() {
+
+        /**********
+         * Persistent storage with claim deletion
+         */
```
Same comment here

scholzj commented:
@tombentley The code with the fixed review comments should be now in.

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java:
```diff
@@ -1520,27 +1530,264 @@ ReconciliationState createReconciliationState(Reconciliation reconciliation, Kaf
+            });
+
+            return futureResult;
+        }
```
the above two methods are quite the same, do you think there is no way for having just one acting on K or ZK provided somehow as parameter?

scholzj replied:
I do not think it is so easy, mainly because the `generatePersistentVolumeClaims` signatures are different for Kafka and for ZK. This is because it already expects that the Kafka storage with JBOD might allow to add or remove volumes (which does not apply for ZK since there is no JBD storage supported). So I would probably end up with something like `instanceof` which would be IMHO quite ugly

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetDiff.java:
```diff
@@ -150,6 +133,11 @@ boolean compareMemoryAndCpuResources(JsonNode source, JsonNode target, String pa
+    /**
+     * Returns whether the Diff is empty or not
+     *
+     * @return
```
use `@return whether the Diff is empty or not`

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java:
```diff
@@ -444,9 +448,29 @@ public Storage getStorage() {
+                    PersistentClaimStorage persistentClaimStorage = (PersistentClaimStorage) jbodVolume;
+                    checkPersistentStorageSize(persistentClaimStorage);
+                }
+            }
```
You would still end up with very similar code, just moved to a different method. Why do you think it would be better there then here? I think it might be more useful here if we need to add other checks in the future, we would not need to go through the JBOD volumes multiple times.

ppatierno replied:
I proposed to move in the checkPersistentStorageSize because what we are doing here is actually just calling .... check on the storage(s).
Maybe if we have more check on storage to do, we could have checkPersistentStorageSize renamed to checkPersistentStorage which contains those checks which for now is only size.

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java:
```diff
@@ -277,53 +276,81 @@ public static String clientsCaCertSecretName(String cluster) {
+            if (!diff.isEmpty()) {
+                log.warn("Changing Kafka storage is not possible. The changes will be ignored.");
+                result.setStorage(oldStorage);
+            } else {
```
I probably prefer the current way TBH.
